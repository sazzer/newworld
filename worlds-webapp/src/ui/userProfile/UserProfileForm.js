// @flow

import React from 'react';
import {Form, Button} from 'semantic-ui-react'
import {Interpolate} from 'react-i18next';
import type {User} from "../../users/users";

/** The flow type representing the props for the User Profile Form */
type UserProfileFormProps = {
    user: User,
    onUpdateUsername: (string) => void,
    onUpdateEmail: (string) => void,
    onUpdateDisplayName: (string) => void,
    onSave: () => void
};

/**
 * Helper to process the onChange event from a control and call a handler with the new value
 */
function processChange(handler) {
    return function(e) {
        handler(e.target.value);
    }
}

/**
 * The form to display for the user profile
 */
export default function UserProfileForm(props: UserProfileFormProps) {
    return (
        <Form>
            <Form.Field>
                <label>Email Address</label>
                <input value={props.user.email}
                       name="email"
                       onChange={processChange(props.onUpdateEmail)}
                       required />
            </Form.Field>
            <Form.Field>
                <label>Username</label>
                <input value={props.user.username}
                       name="username"
                       onChange={processChange(props.onUpdateUsername)}
                       required />
            </Form.Field>
            <Form.Field>
                <label>Display Name</label>
                <input value={props.user.displayName}
                       name="displayName"
                       onChange={processChange(props.onUpdateDisplayName)}
                       required />
            </Form.Field>
            <Button primary onClick={props.onSave}>Save</Button>
        </Form>
    );
}
