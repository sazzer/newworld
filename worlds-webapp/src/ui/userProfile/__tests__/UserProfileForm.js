// @flow

import React from 'react';
import {shallow} from 'enzyme';
import UserProfileForm  from '../UserProfileForm';

/** Create the component to test */
function setup() {
    const user = {
        id: 'id',
        displayName: 'Test User',
        username: 'testuser'
    };

    const onUpdateEmail = jest.fn();
    const onUpdateUsername = jest.fn();
    const onUpdateDisplayName = jest.fn();
    const onSave = jest.fn();

    return {
        user,
        onUpdateEmail,
        onUpdateUsername,
        onUpdateDisplayName,
        onSave,
        element: shallow(<UserProfileForm user={user}
                                          onUpdateEmail={onUpdateEmail}
                                          onUpdateUsername={onUpdateUsername}
                                          onUpdateDisplayName={onUpdateDisplayName}
                                          onSave={onSave} />)
    };
}

describe('UserProfileForm', () => {
    it('Renders correctly', () => {
        const { element } = setup();

        expect(element).toMatchSnapshot();
    });

    it('Triggers onSave when the save button is clicked', () => {
        const { element, onSave, onUpdateEmail, onUpdateUsername, onUpdateDisplayName } = setup();

        element.find("Button").simulate("click");

        expect(onSave).toHaveBeenCalledTimes(1);

        expect(onUpdateEmail).toHaveBeenCalledTimes(0);
        expect(onUpdateUsername).toHaveBeenCalledTimes(0);
        expect(onUpdateDisplayName).toHaveBeenCalledTimes(0);
    });

    it('Triggers onUpdateEmail when the email address is changed', () => {
        const { element, onSave, onUpdateEmail, onUpdateUsername, onUpdateDisplayName } = setup();

        element.find("input[name='email']").simulate('change', {
            target: {
                value: 'new@example.com'
            }
        });

        expect(onUpdateEmail).toHaveBeenCalledTimes(1);
        expect(onUpdateEmail).toHaveBeenCalledWith('new@example.com');

        expect(onUpdateUsername).toHaveBeenCalledTimes(0);
        expect(onUpdateDisplayName).toHaveBeenCalledTimes(0);
        expect(onSave).toHaveBeenCalledTimes(0);
    });

    it('Triggers onUpdateUsername when the username is changed', () => {
        const { element, onSave, onUpdateEmail, onUpdateUsername, onUpdateDisplayName } = setup();

        element.find("input[name='username']").simulate('change', {
            target: {
                value: 'newUser'
            }
        });

        expect(onUpdateUsername).toHaveBeenCalledTimes(1);
        expect(onUpdateUsername).toHaveBeenCalledWith('newUser');

        expect(onUpdateEmail).toHaveBeenCalledTimes(0);
        expect(onUpdateDisplayName).toHaveBeenCalledTimes(0);
        expect(onSave).toHaveBeenCalledTimes(0);
    });

    it('Triggers onUpdateDisplayName when the display name is changed', () => {
        const { element, onSave, onUpdateEmail, onUpdateUsername, onUpdateDisplayName } = setup();

        element.find("input[name='displayName']").simulate('change', {
            target: {
                value: 'New User'
            }
        });

        expect(onUpdateDisplayName).toHaveBeenCalledTimes(1);
        expect(onUpdateDisplayName).toHaveBeenCalledWith('New User');

        expect(onUpdateEmail).toHaveBeenCalledTimes(0);
        expect(onUpdateUsername).toHaveBeenCalledTimes(0);
        expect(onSave).toHaveBeenCalledTimes(0);
    });
});
