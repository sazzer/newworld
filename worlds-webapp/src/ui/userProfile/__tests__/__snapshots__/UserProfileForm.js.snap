// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileForm Renders correctly 1`] = `
<Form
  as="form"
>
  <FormField>
    <label>
      Email Address
    </label>
    <input
      name="email"
      onChange={[Function]}
      required={true}
    />
  </FormField>
  <FormField>
    <label>
      Username
    </label>
    <input
      name="username"
      onChange={[Function]}
      required={true}
      value="testuser"
    />
  </FormField>
  <FormField>
    <label>
      Display Name
    </label>
    <input
      name="displayName"
      onChange={[Function]}
      required={true}
      value="Test User"
    />
  </FormField>
  <Button
    as="button"
    onClick={[Function]}
    primary={true}
    role="button"
  >
    Save
  </Button>
</Form>
`;
